### PR TITLE
Version 10Feb21

### PR DESCRIPTION
**Pull request by alpakpinar**
Hey @AndreasAlbert , before more changes happen, I wanted to open this PR as a reference for the current version of the code I'm using for the studies. Here are the main updates in this PR:

- Multiprocessing: Dividing the input file into chunks and parallel processing (number of parallel jobs also being configurable)
- File I/O: We're saving one ROOT file per event chunk, workspaces being saved inside the file as usual.
- Added a new script under plot/ to extract HTmiss distributions before and after from the workspace, and plots them
- Dummy JER option: I was studying the impact of using placeholder Gaussians as response, so I added it as a running option  (it's a bit of a hack right now, I can update it if you want to, or just remove it from PR if we don't want it in master?)

Let me know if you want any changes, and I can work on those updates, thanks!

### CODE REVIEW

File: rebalance.py
Comment (AndreasAlbert) [inline rebalance.py]: I think this implementation is a bit suboptimal because you touch code in many places to accomplish one small thing you are trying to check. We should keep independent logical units (unit 1: JER, unit 2: workspace creation) apart as far as possible. The changes introduced here do not keep them apart.

You could make this much easier by making the JERLookup class more flexible and extracting its construction from the WS factory. A way to accomplish this would be to give class methods to `JERLookup` that initiate the object based on different inputs:

```python
jer1 = JerLookUp.from_th1(some_th1)
jer2 = JerLookUp.from_constant(constant_resolution_value)
```

After being constructed, the usage would still be identical:

```python
jer1.get_jer(pt, eta) # changes depending on arguments
jer2.get_jer(pt, eta) # always returns the same value 
```

For the workspace factory, we can then replace the `set_jer_source` method by `set_jer_evaluator` and pass in a pre-initiated JerLookup object. 

```python
jer = JerLookUp.from_th1(some_th1)
wsfactory.set_jer_source(jer)
```


That way, we accomplish two things:

1. Most of the ws factory does not have to change to accommodate our change to resolution.
2. If we ever want to implement additional ways of evaluating JER (one could think to use fit functions, or whatever else), we do not need to touch the ws factory at all. Ever. As long as we can write our resolution input down in a way that makes it evaluable based on (pt, eta), we can just add extensions to JerLookup.

File: exec.py
Comment (AndreasAlbert) [inline exec.py]: bucoffea is a relatively volatile package and adds a huge dependency chain to this otherwise lightweight project. Not worth it just in order to get two small utility functions. Just copy the implementation into utility functions in this package.

File: exec.py
Comment (AndreasAlbert) [inline exec.py]: For clarity, I don't think a fcuntion called `read_jets` without cut-related arguments should include hard-coded jet cuts (not very clear to the user). Instead, I'd suggest to either:

1. Remove all cuts from this function. User gets a list of all jets, and then has to cut themselves ```jets = [j for j in read_jets(event) if j.pt > whatever and abs(j.eta)<whatever]```
2. add defined arguments `ptmin`, `absetamax` to the function signature and use those to deifne cuts.

File: exec.py
Comment (AndreasAlbert) [inline exec.py]: I'd suggest to remove all plotting from this script. The plane plots were a nice per-event debugging tool to get this fit started, but if you are doing serious processing, you are not going to look at 10^N plots. Suggest to move all plotting into a separate script so that it is still easy to make plots for a given workspace if needed.

File: exec.py
Comment (AndreasAlbert) [inline exec.py:47]: seems like there's something missing here

File: exec.py
Comment (AndreasAlbert) [inline exec.py]: I'm not familiar with this construction. Don't you want to use `multiprocessing.Pool` here?
Your code seems to indicate that the number of chunks is always identical to the number of parallel jobs to start. That implies that the chunks can get very big if the input is very big. Wouldn't it make sense to decouple the two so that you can enforce ~constant chunk SIZES rather than chunk COUNTS?

**Review by AndreasAlbert — CHANGES_REQUESTED** [review]

File: exec.py
Comment (AndreasAlbert) [reply exec.py]: even more so given that the functions in question [are literal one-liners](https://github.com/bu-cms/bucoffea/blob/master/bucoffea/helpers/git.py#L5-L9)

**Comment by alpakpinar** [comment]
Thanks a lot for the detailed comments @AndreasAlbert! Will work on updating the PR according to your comments and I'll let you know when I have a new version, thanks

File: exec.py
Comment (alpakpinar) [reply exec.py:47]: What do you think is missing here? Chunksize is also embedded into the args, if you were thinking about that. Or is there anything else you were thinking about, thank you!

File: exec.py
Comment (alpakpinar) [reply exec.py]: Thanks for the suggestion, indeed ``multiprocessing.Pool`` works very nicely here. In 636d195, I replaced the earlier construction with the ``Pool``object. This way, as you said, It works nicely as we divide a large input into smaller chunks.

File: rebalance.py
Comment (alpakpinar) [reply rebalance.py]: Thanks for the detailed suggestions! In 117e041, I tried to do a similar implementation as you outlined. Let me know if it looks good or you want modifications to this, thank you!

File: exec.py
Comment (AndreasAlbert) [reply exec.py:47]: Sorry, I got confused by the way github shows this part of the diff. When I looked at it before, it seemed to me like the function ended after the `infile` line, i.e. before doing anything. In reality, it's fine.

File: rebalance.py
Comment (AndreasAlbert) [reply rebalance.py]: thanks look good

**Comment by AndreasAlbert** [comment]
Looks good to me now. Shall I merge?

**Review by AndreasAlbert — APPROVED** [review]

**Comment by alpakpinar** [comment]
Sure, this should be ready to merge if it looks good to you, thanks!

**Pull request merged**
